### PR TITLE
Adding ligand processing to dataset and encoders.

### PR DESCRIPTION
**Pull request by vratins**
Adds non-protein, non-water heavy atoms as context nodes, on by default (`include_ligands=True`).

  - **Dataset:** `parse_asu_with_biotite()` returns `(protein, water, ligand)`; ligand atoms appended after protein/mate atoms with an `is_ligand` mask and `residue_index = -1`. `include_ligands` set to false leads to the cache name having a `_noligands` suffix.
  - **Encoder:** `CachedEmbeddingEncoder`: each protein-type node is described by two modalities — its cached ESM/SLAE embedding (zero-padded for ligands and mates, which have no residue embedding) and its per-atom element one-hot. Both are projected to fusion_dim, normalized per stream, and fused through a small MLP.
  - **Misc:** ESM/SLAE scripts unpack the 3-tuple (SLAE marked legacy); removed unused `rdkit`; added ligand tests + `4h0b` fixture.
  - Added fusion MLP to the encoders to make the logic cleaner (ESM and 1hots projected to common dim, and then concat, and another MLP, and passed to flow.

## Summary by CodeRabbit

* **New Features**
  * Added support for including ligand atoms from PDB structures in model training
  * Introduced configurable toggle to enable/disable ligand atom inclusion

* **Changes**
  * Encoder initialization now requires explicit embedding dimension specification at configuration time
  * Updated dependencies for improved compatibility

* **Tests**
  * Expanded test coverage for ligand processing and atom filtering

### CODE REVIEW

File: src/dataset.py [hunk 1]
```diff
@@ -1164,6 +1214,7 @@ def __getitem__(self, idx: int) -> HeteroData:
         protein_pos = cached["protein_pos"]
         protein_x = cached["protein_x"]
         protein_res_idx = cached["protein_res_idx"]
+        is_ligand = cached["is_ligand"]
         pp_edge_index = cached["pp_edge_index"]
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:1273]: Backward compatibility issues, not concerned here, ligands should be processed by default.

File: src/dataset.py [hunk 2]
```diff
@@ -720,7 +738,9 @@ def __init__(
         """
 
         self.cache_dir = Path(processed_dir)
-        # Directory-based separation: geometry/ vs geometry_mates/
+        # Directory-based separation: geometry/ vs geometry_mates/. Ligand inclusion
+        # is governed by the include_ligands config flag, not the cache directory
+        # name, so the geometry cache name is unaffected by include_ligands.
         cache_suffix = "_mates" if include_mates else ""
         self.geometry_dir = self.cache_dir / f"{geometry_cache_name}{cache_suffix}"
         self.base_pdb_dir = Path(base_pdb_dir)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:777]: include_ligands should be on by default, not a concern here, both caches should have ligands

File: src/dataset.py [hunk 3]
```diff
@@ -1046,6 +1069,32 @@ def _preprocess_one(self, entry: dict, cache_path: Path):
...
             final_protein_res_idx = protein_res_idx
 
+        # Append ligand atoms after protein (and mate) atoms when enabled.
+        # is_ligand mask marks which protein-type nodes are ligand atoms.
+        # Ligands always go last so num_asu_protein and mate counts are unaffected,
+        # preserving ESM/SLAE embedding alignment via _pad_atom_embeddings_for_mates.
+
+        # TODO(ligands+mates): this only adds ASU ligands. Until dev_crystal_mates
+        # is opened for a PR, mates are restricted to polymer.protein, so a ligand sitting in a
+        # crystal contact is dropped from the mates
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py]: Future PR

File: src/encoder_base.py [hunk 4]
```diff
@@ -140,57 +142,73 @@ def from_config(cls, config: dict, device: torch.device) -> BaseProteinEncoder:
...
+        self._embedding_dim: int = embedding_dim
+        self._fusion_dim: int = fusion_dim
         self._embedding_key = embedding_key
         self._encoder_type = encoder_type
+        # Project each modality to fusion_dim, LayerNorm each stream, then fuse.
+        # Separate projections + per-stream norm keep the 16-dim element signal from
+        # being swamped by the wide, large-norm ESM vector (raw ESM norms are ~1e3-1e4
+        # vs 1 for a one-hot), so both modalities enter the fuse MLP at comparable scale.
+        self.esm_proj = nn.Linear(embedding_dim, fusion_dim)
+        self.elem_proj = nn.Linear(NODE_FEATURE_DIM, fusion_dim)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/encoder_base.py:202]: SLAE is legacy, ESM will remain primary encoder for now.

File: src/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DorisMai) [reply src/dataset.py:777]: Why is this not a concern? I understand that the default is include_ligands=True so for your own caches this is fine, but if another user wants to toggle and try excluding ligands would there not be silent errors from loading caches?

File: src/dataset.py [hunk 5]
```diff
@@ -665,6 +675,7 @@ def __init__(
         base_pdb_dir: str = "/sb/wankowicz_lab/data/srivasv/pdb_redo_data",
```
Comment (DorisMai) [inline src/dataset.py]: out side of the diff, but just hard-coded path to remove

File: src/dataset.py
Comment (DorisMai) [inline src/dataset.py:913]: this function is getting very very long...not really a blocking comment for this PR but if there is a future PR for cleaning up / refactoring this should be on the top of the list.

File: src/dataset.py [hunk 6]
```diff
@@ -1046,6 +1069,32 @@ def _preprocess_one(self, entry: dict, cache_path: Path):
             final_protein_x = protein_x
             final_protein_res_idx = protein_res_idx
 
+        # Append ligand atoms after protein (and mate) atoms when enabled.
+        # is_ligand mask marks which protein-type nodes are ligand atoms.
+        # Ligands always go last so num_asu_protein and mate counts are unaffected,
+        # preserving ESM/SLAE embedding alignment via _pad_atom_embeddings_for_mates.
+
+        # TODO(ligands+mates): this only adds ASU ligands. Until dev_crystal_mates
```
Comment (DorisMai) [inline src/dataset.py]: I am a bit confused by the comment. Is the intention (future PR) to add ligands as part of mate protein nodes or not?

File: tests/test_dataset.py [hunk 7]
```diff
@@ -480,32 +480,223 @@ class TestParseAsuWithBiotite:
     """Tests for PDB parsing with biotite."""
 
     def test_parse_returns_protein_and_water(self, pdb_6eey):
```
Comment (DorisMai) [inline tests/test_dataset.py]: should rename this function

File: tests/test_encoder.py [hunk 8]
```diff
@@ -142,23 +144,33 @@ def test_build_encoder_slae(self, device):
...
-        # output_dims not available until forward pass
+        # output_dims is the fused width (= hidden_s), not the cached embedding_dim
+        assert encoder.output_dims == (64, 0)
 
     def test_build_encoder_esm(self, device):
         """build_encoder should construct ESM encoder from config."""
         config = {
             "encoder_type": "esm",
+            "embedding_dim": 1536,
+            "hidden_s": 64,
```
Comment (DorisMai) [inline tests/test_encoder.py]: where do these hard coded numbers come from (and similarly 128 above)? should some constants be declared/defined?

File: src/encoder_base.py [hunk 9]
```diff
@@ -140,57 +142,73 @@ def from_config(cls, config: dict, device: torch.device) -> BaseProteinEncoder:
...
-    This pass-through encoder reads embeddings stored in HeteroData under a
-    specified key and returns them as scalar features. No neural network
-    computation occurs; all geometric processing happens in downstream layers.
+    Fusion encoder for pre-computed protein embeddings (ESM, SLAE, etc.).
+
+    Each protein-type node is described by two modalities, both already present
+    on data['protein']:
+      - a cached sequence/structure embedding (`embedding`, width embedding_dim).
+        ASU protein atoms carry the real ESM/SLAE vector; symmetry mates and
+        ligand atoms are zero-padded (they have no residue embedding).
```
Comment (DorisMai) [inline src/encoder_base.py:151]: I am not seeing relevant tests for testing if the embeddings work as expected -- whether should be zero-padded or not and have the one-hot component etc for ASU protein, mate, and ligands? I know the mate is not part of this PR, but probably should start adding tests to catch the kind of bug that you found that prevented crystal contact to work.

**Comment by DorisMai** [comment]
I made a few comments that should mostly be minor issues to address or clarifying questions related to other PRs perhaps. Other than that, README.md should also be updated wherever ligand should be mentioned as optionally a part of "protein" now.

**Review by DorisMai — CHANGES_REQUESTED** [review]
Main blocking issue I am requesting to address is add test, update README to reflect the code changes, and clarify cache name assumption.

File: src/dataset.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:777]: Fair concern honestly. I'll add a no_ligands suffix to the cache name in the case someone wants to generate the cache with no ligands, with the falg being default true.

File: src/dataset.py (hunk 5 above)
Comment (vratins) [reply src/dataset.py]: Done

File: src/dataset.py
Comment (vratins) [reply src/dataset.py:913]: Adding an issue to clean up preprocessing method in src/dataset.py and break it up.

File: src/dataset.py (hunk 6 above)
Comment (vratins) [reply src/dataset.py]: Added clarity to the comment. It will be part of the PR.

File: tests/test_dataset.py (hunk 7 above)
Comment (vratins) [reply tests/test_dataset.py]: Done to reflect 3 return types now.

File: tests/test_encoder.py (hunk 8 above)
Comment (vratins) [reply tests/test_encoder.py]: Declared the dimensions as constant.

File: src/encoder_base.py (hunk 9 above)
Comment (vratins) [reply src/encoder_base.py:151]: I'll add these tests comprehensively to the dev_crystal_mates PR.

File: src/dataset.py [hunk 10]
```diff
@@ -749,8 +767,12 @@ def __init__(
...
 
         self.cache_dir = Path(processed_dir)
-        # Directory-based separation: geometry/ vs geometry_mates/
+        # Directory-based separation: geometry/ vs geometry_mates/. Both flags change
+        # the cached node set, so both are encoded in the directory name -- otherwise
+        # toggling one would silently reuse geometry built under the other setting.
         cache_suffix = "_mates" if include_mates else ""
+        if not include_ligands:
+            cache_suffix += "_noligands"
         self.geometry_dir = self.cache_dir / f"{geometry_cache_name}{cache_suffix}"
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:776]: Done

File: src/encoder_base.py [hunk 11]
```diff
@@ -140,57 +142,73 @@ def from_config(cls, config: dict, device: torch.device) -> BaseProteinEncoder:
...
-        self._embedding_dim: int | None = embedding_dim
+        self._embedding_dim: int = embedding_dim
+        self._fusion_dim: int = fusion_dim
         self._embedding_key = embedding_key
         self._encoder_type = encoder_type
+        # Project each modality to fusion_dim, LayerNorm each stream, then fuse.
+        # Separate projections + per-stream norm keep the 16-dim element signal from
+        # being swamped by the wide, large-norm ESM vector (raw ESM norms are ~1e3-1e4
+        # vs 1 for a one-hot), so both modalities enter the fuse MLP at comparable scale.
+        self.esm_proj = nn.Linear(embedding_dim, fusion_dim)
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/encoder_base.py:201]: Clarified in the docstring.

File: src/dataset.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:776]: Done

File: src/dataset.py [hunk 12]
```diff
@@ -1079,6 +1104,33 @@ def _preprocess_one(self, entry: dict, cache_path: Path):
...
 
+        # Append ASU ligand atoms after protein (and mate) atoms when enabled.
+        # is_ligand mask marks which protein-type nodes are ligand atoms.
+        # Ligands always go last so num_asu_protein and mate counts are unaffected,
+        # preserving ESM/SLAE embedding alignment via _pad_atom_embeddings_for_mates.
+        #
+        # TODO(ligands+mates): mate ligands should be included as protein-type nodes
+        # too, but are not yet. Mate generation is restricted to polymer.protein, so a
+        # ligand sitting in a crystal contact is currently dropped. Extending mate
+        # generation to het atoms is deferred to dev_crystal_mates.
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py]: Will update the docstring with a todo to fix the mates bugs.

**Review by vratins — COMMENTED** [review]
Addressed comments, and added docstring comments for what is left for future PRs.

File: src/encoder_base.py [hunk 13]
```diff
@@ -221,12 +241,23 @@ def forward(
...
+                f"{self._encoder_type.upper()} encoder configured with "
+                f"embedding_dim={self._embedding_dim}, but cached "
+                f"'{self._embedding_key}' embeddings have width {embeddings.size(-1)}. "
+                f"Ensure the encoder config matches the cached embeddings."
+            )
+
+        x = data["protein"].x.to(embeddings.device)
+        esm = self.esm_norm(self.esm_proj(embeddings))
+        elem = self.elem_norm(self.elem_proj(x))
+        fused = self.fuse(torch.cat([esm, elem], dim=-1))
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/encoder_base.py]: Added checks and tests towards this.

File: src/encoder_base.py [hunk 14]
```diff
@@ -140,57 +142,76 @@ def from_config(cls, config: dict, device: torch.device) -> BaseProteinEncoder:
...
-    specified key and returns them as scalar features. No neural network
-    computation occurs; all geometric processing happens in downstream layers.
+    Fusion encoder for pre-computed protein embeddings (ESM, SLAE, etc.).
+
+    Each protein-type node is described by two modalities, both already present
+    on data['protein']:
+      - a cached sequence/structure embedding (`embedding`, width embedding_dim).
+        ASU protein atoms carry the real ESM/SLAE vector; symmetry mates and
+        ligand atoms are zero-padded (they have no residue embedding).
+      - a per-atom element one-hot (`x`, width NODE_FEATURE_DIM).
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/encoder_base.py:152]: Yes ligand_embed is legacy, fixed this.

File: README.md [hunk 15]
```diff
@@ -100,17 +102,20 @@ Preprocessed data is cached under `--processed_dir` in a three-layer architectur
...
 │       - num_asu_protein: ASU atom count (mate boundary metadata)
 │       # Note: When include_mates=True, mate atoms are concatenated into
-│       # protein_pos/protein_x. Recover boundaries via:
-│       #   ASU atoms = protein_pos[:num_asu_protein]
-│       #   Mate atoms = protein_pos[num_asu_protein:]
+│       # protein_pos/protein_x, and ligand atoms are appended after those.
+│       # Node order is [ASU protein | mates | ligands]. Recover boundaries via:
+│       #   ASU atoms    = protein_pos[:num_asu_protein]
+│       #   Ligand atoms = protein_pos[is_ligand]              # always last
+│       #   Mate atoms   = protein_pos[num_asu_protein:][~is_ligand[num_asu_protein:]]
```
[the remark this answers is not in the document]
Comment (vratins) [reply README.md]: Done

File: src/dataset.py [hunk 16]
```diff
@@ -678,11 +688,12 @@ def __init__(
         self,
         pdb_list_file: str,
         processed_dir: str,
+        base_pdb_dir: str,
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/dataset.py:691]: Done
